server/game/gameSystems/DiscardHandSystem.ts

```diff
@@ -0,0 +1,32 @@
+import { EventName, Location } from '../core/Constants';
+import { AbilityContext } from '../core/ability/AbilityContext';
+import * as Helpers from '../core/utils/Helpers.js';
+import type Player from '../core/Player';
+import { IPlayerTargetSystemProperties, PlayerTargetSystem } from '../core/gameSystem/PlayerTargetSystem';
+
+// eslint-disable-next-line @typescript-eslint/no-empty-object-type
+export interface IDiscardHandSystemProperties extends IPlayerTargetSystemProperties {}
+
+/**
+ * A {@link GameSystem} which discards a player's entire hand
+ */
+export class DiscardHandSystem<TContext extends AbilityContext = AbilityContext> extends PlayerTargetSystem<TContext, IDiscardHandSystemProperties> {
+    public override name = 'discardHand';
+    public override eventName = EventName.OnHandDiscarded;
+    public override readonly costDescription: string = 'discard hand';
+    public override readonly effectDescription: string = 'discard hand';
+
+    public override eventHandler(event): void {
+        const players = Helpers.asArray(event.player) as Player[];
+
+        players.forEach((player) =>
+            player
+                .getCardPile(Location.Hand)
+                .forEach((card) => player.moveCard(card, Location.Discard))
+        );
+    }
+
+    public override defaultTargets(context: TContext): Player[] {
```

Comment:
I'd rather not override the default target behavior unless it's on a system where it is super clear that the natural default should be different, just so people don't get tripped up on needing to check what the default is for each system

Reply:
The problem is that there isn't an easy way to pass a player to the system. This, by default, target the opponent player because `PlayerTargetSystem` is using that:
```Typescript
immediateEffect: AbilityHelper.immediateEffects.discardEntireHand()
```

So you would need to do something like this:
```Typescript
immediateEffect: AbilityHelper.immediateEffects.discardEntireHand({ target: context.player })
```

But you can't because you don't have a context.

The alternatives would be:
1. To add a way to set `immediateEffect` with a function that takes a `context`.
2. To add a way to set up a player target resolver where you can return the player you want.

I guess (2) would be the best one?

Reply:
Oh yeah so (1) is already implemented in the engine. Every GameSystem (immediateEffect) can accept either a properties object or a handler that constructs one from a context object. If the latter, then the engine will automatically call it to create the properties object at the appropriate time with the right context object.

This works for both card target and player target systems - see Smuggling Compartment for an example of the latter.

Reply:
Of course it does, I already use that many times!

 For some reason my brain was trying to do:
```Typescript
(context) => AbilityHelper.immediateEffects.discardEntireHand({ target: context.player })
```

Instead of:
```Typescript
AbilityHelper.immediateEffects.discardEntireHand((context) => ({ target: context.player }))
```